plugins/parcelize/parcelize-compiler/testData/diagnostics/customParcelers.fir.kt

```diff
@@ -19,27 +19,27 @@ class StringClassParceler : Parceler<String> {
     override fun String.write(parcel: Parcel, flags: Int) = TODO()
 }
 
-@TypeParceler<String, StringParceler>
-class MissingParcelizeAnnotation(val a: @WriteWith<StringParceler> String)
+@<!CLASS_SHOULD_BE_PARCELIZE!>TypeParceler<!><String, StringParceler>
+class MissingParcelizeAnnotation(val a: @<!CLASS_SHOULD_BE_PARCELIZE!>WriteWith<!><StringParceler> String)
 
 @Parcelize
 @TypeParceler<String, StringClassParceler>
-class ShouldBeClass(val a: @WriteWith<StringClassParceler> String) : Parcelable
+class ShouldBeClass(val a: @WriteWith<<!PARCELER_SHOULD_BE_OBJECT!>StringClassParceler<!>> String) : Parcelable
 
 @Parcelize
 class Test(
-    val a: @WriteWith<StringParceler> Int,
+    val a: @WriteWith<<!PARCELER_TYPE_INCOMPATIBLE!>StringParceler<!>> Int,
     val b: @WriteWith<StringParceler> String,
-    val c: @WriteWith<StringParceler> CharSequence,
-    val d: @WriteWith<CharSequenceParceler> String,
+    val c: @WriteWith<<!PARCELER_TYPE_INCOMPATIBLE!>StringParceler<!>> CharSequence,
+    val d: @WriteWith<<!PARCELER_TYPE_INCOMPATIBLE!>CharSequenceParceler<!>> String,
```

Comment:
Is it ok that K1 implementation does not report `PARCELER_TYPE_INCOMPATIBLE` on `d`?

Reply:
It's going to be a separate PR/issue, since the check in K1 for `WriteWith` is wrong, imo. Or rather, it is explicitly checking that the given parceler can be used to *write* the annotated type, but the generated code will also use it to try and read the same type.

More precisely, in K1 we handle `@WriteWith<P> T1` by finding a `Parceler<T2>` superclass of `P` and then testing that `T1` is a subtype of `T2`. In this example that succeeds for `d` since `T2` is `CharSequence` and `T1` is `String`. If instead we want to check that we can both read and write the annotated type we should check that `P` is a subtype of `Parceler<T1>`, which is what the FIR checker does.

Bottom line is I think the existing check is wrong and I didn't want to reproduce the buggy behavior in K2. I'll change it to match if it turns out that I'm wrong and that there's a legitimate use case here...

---

Incidentally, the same problem exists for `TypeParceler`, but that's just in the form of a variance annotation in the definition of the annotation.

Reply:
Got it, thanks. In that case, I'll merge PR after CI build will pass